// @metano/quasar/stores/AuthStore.js
import { defineStore } from 'pinia'
import { getStorage, setStorage } from '@metano/quasar/boot/storage'
import { HTTPAuth, HTTPClient, url } from '@metano/quasar/boot/api'



export const LoadStore = defineStore('load', {
  state: () => ({
    count: 0
  }),
  getters: {
    value: (state) => state.counter,
  },
  actions: {
    inc() {
      this.count++
    },
    dec() {
      this.count--
    },
  }
})


export const AlertStore = defineStore('alert', {
  state: () => ({
    data: []
  }),
  actions: {
    add(alert) {
      this.data.push(alert);
    },
    remove(alert) {
      this.data = this.data.filter(a => a.id !== alert.id);
    },
    clear() {
      this.data = []
    },
  }
})


export const UserStore = defineStore("user", {
  state: () => ({
    data: null,
    Entidades: [{id: '1',  nome: 'Mytech' }],
    Entidade: {id: '1',  nome: 'Mytech' },
    Sucursals: [{ id: '1', nome: 'Sede Mytech' }],
    Sucursal: { id: '1', nome: 'Sede Mytech' },
    Grupos: [],
    Grupo: {id: '1',  name: 'Gest' },
    Idioma: {id: 1, nome: "Português", code: "PT-PT"} ,
    Idiomas: [
      { id: 1, nome: "Português", code: "PT-PT" },
      { id: 2,  nome: "English", code: "EN" },
    ],
    Settings: false,
    Permicoes: [],
    access: null,
    refresh: null,
    LeftTop: true,
    LeftMenu: true,
  }),

  getters: {
    username: (state) => state.data?.username || ".",
    perfil: (state) =>
      state.data?.perfil ||
      "https://cdn.quasar.dev/img/boy-avatar.png",
  },

  actions: {
    setIdioma(idioma){
      this.Idioma = idioma
    },
    selectGroup(grupo){
      this.Grupo = grupo
    },
    toggleSettings(){
      this.Settings = !this.Settings
      setStorage('c', 'settings', this.Settings)
    },
    toggleLeftTop(){
      this.LeftTop = !this.LeftTop
      setStorage('c', 'left_top', this.LeftTop)
    },
    toggleLeftMenu(){
      this.LeftMenu = !this.LeftMenu
      setStorage('c', 'left_menu', this.LeftMenu)
    },

    async login(data) {
      await HTTPClient.post(url({type: "u", url: "auth/login/", params: {}}), data )
      .then(res => {
        this.access = res.data.tokens.access
        this.refresh = res.data.tokens.refresh
        setStorage('c', 'access', this.access,  365)
        setStorage('c', 'refresh', this.refresh,  365)
        this.me()
      }).catch(err => {
        console.log(err)
      })
    },

    async me() {
      await HTTPAuth.get(url({type: "u", url: "auth/me", params: {}}) )
      .then(res => {
        this.data = res.data
        setStorage('c', 'user', JSON.stringify(this.data),  365)
      }).catch(err => {
        console.log(err)
      })
    },


    async refreshToken() {
      const data = {refresh: this.refresh }
      await HTTPAuth.post(url({type: "u", url: "auth/refresh_token/", params: {}}), data )
      .then(res => {
        this.access = res.data.access
      }).catch(err => {

      })
    },

    async change_password_email(email, antiga, nova) {
      const data = { email: email, password: antiga, passwordNova: nova }
      await HTTPAuth.post(url({type: "u", url: "auth/change_password_email/", params: {}}), data )
      .then(res => {
      }).catch(err => {

      })
    },

    async change_password_numero(mobile, otp, nova) {
      const data = { mobile: mobile, otp: otp, password: nova }
      await HTTPAuth.post(url({type: "u", url: "auth/change_password_email/", params: {}}), data )
      .then(res => {
      }).catch(err => {

      })
    },


    async logout(x) {
      
      await HTTPAuth.post(url({type: "u", url: "auth/logout/", params: {}}), {refresh: this.refresh} )
      .then(res => {
        this.data = null;
        this.refresh = null;
        this.access = null;
        this.Grupos = [];
        this.Grupo = null;

        const userEntidade = getStorage('c', 'userEntidade')
        const entidade = getStorage('c', 'entidade')
        
        deleteStorage('c', 'access')
        deleteStorage('c', 'refresh')
        deleteStorage('c', 'user')
        deleteStorage('c', 'userGrupo')
        deleteStorage('c', 'userEntidade')
        deleteStorage('c', 'entidade')
        deleteStorage('c', 'userGrupo')
        deleteStorage('c', 'userEntidades')
        deleteStorage('c', 'userSucursals')
        deleteStorage('c', 'userSucursal')
        deleteStorage('c', 'userPessoa')
        deleteStorage('c', 'userPerfils')
        deleteStorage('c', 'linga')
        deleteStorage('c', 'entidadeModulos')

        deleteStorage('l', 'traducao')
        deleteStorage('l', 'userPermicoes')
        deleteStorage('l', 'manterlogado')
        deleteStorage('l', 'username')
        deleteStorage('l', 'password')

        if (x === 'x') {
          this.$router.push({ name: 'login' })
        } else {
          setStorage('c', 'userEntidade', userEntidade, 365)
          setStorage('c', 'entidade', entidade, 365)
          this.$router.push({ name: 'login_entidade', params: { entidade_id: x } })
        }

      }).catch(err => {

      })
    },

    isAuthorized (permission) {
      let result = false
      if (Array.isArray(this.Permicoes)) {
        this.Permicoes.forEach(element => {
          if (element.nome === permission) {
            result = true
          } else {

          }
        })
      }
      return result
    },

  },
});


export const AuthStore = defineStore('auth', {
  state: () => ({
    TipoEntidades: [{ id: 'ed174ebc-c8b6-4c0b-b950-58cc8c4a4e26', nome: 'Mytech' }, { id: 'ed174ebc-c8b6-4c0b-b950-58cc8c4a4e26', nome: 'Clinica' }],
    TipoEntidade: { id: 'ed174ebc-c8b6-4c0b-b950-58cc8c4a4e26', nome: 'Mytech' },
    Idiomas: [ ]
  }),

  getters: {

  },

  actions: {
    async getTipoEntidades(){
      await HTTPClient.get(url({type: "u", url: "auth/tipoEntidades", params: {}}) )
      .then(res => {
        this.TipoEntidades = res.data
      }).catch(err => {

      })
    },
  }
})

export const LanguageStore = defineStore("lang", {
  state: () => ({
    current: {id: 1, nome: "Português", code: "PT-PT"} ,
    list: [
      { id: 1, nome: "Português", code: "PT-PT" },
      { id: 2,  nome: "English", code: "EN" },
      { id: 3, nome: "Français", code: "FR" },
    ],
  }),

  actions: {
    change(lang) {
      this.current = lang;
    },

    async get() {
      await HTTPClient.get(url({type: "u", url: "auth/idiomas", params: {}}) )
      .then(res => {
        this.list = res.data
      }).catch(err => {

      })

    },
  },
});
